crates/revm/src/evm_impl.rs

```diff
@@ -99,6 +99,11 @@ impl<'a, GSPEC: Spec, DB: Database, const INSPECT: bool> Transact<DB::Error>
             return Err(InvalidTransaction::RejectCallerWithCode.into());
         }
 
+        // Check that the transaction's chain id is correct
+        if self.data.env.cfg.chain_id != U256::from(self.data.env.tx.chain_id.unwrap_or_default()) {
```

Comment:
There are legacy transactions that does not have a chain_id. It is optional field and if  tx does not have it we don't need to check it. do `if let Some(tx_chain_id) = self.data.env.tx.chain_id {`